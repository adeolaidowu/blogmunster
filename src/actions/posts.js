import database from "../firebase/firebase";

//ADD_POST
export const addPost = (post) => ({
  type: "ADD_POST",
  post,
});

export const startAddPost = (postData = {}) => {
  return (dispatch) => {
    const {
      title = "",
      content = "",
      imageLink = "",
      createdAt = 0,
    } = postData;
    const post = { title, content, imageLink, createdAt };
    return database
      .ref("posts")
      .push(post)
      .then((ref) => {
        dispatch(
          addPost({
            id: ref.key,
            ...post,
          })
        );
      });
  };
};

//EDIT_POST
export const editPost = (id, updates) => ({
  type: "EDIT_POST",
  id,
  updates,
});

export const startEditPost = (id, updates) => {
  return (dispatch) => {
    dispatch(editPost(id, updates));
    database.ref(`posts/${id}`).update(updates);
  };
};

//DELETE_POST
export const deletePost = ({ id } = {}) => ({
  type: "DELETE_POST",
  id,
});

export const startDeletePost = ({ id } = {}) => {
  return (dispatch) => {
    database
      .ref(`posts/${id}`)
      .remove()
      .then(() => {
        dispatch(deletePost({ id }));
      });
  };
};

//SET_POSTS -- This action fetches data from db to display when page is visited
export const setPosts = (posts) => ({
  type: "SET_POSTS",
  posts,
});

export const startSetPosts = () => {
  return (dispatch) => {
    return database
      .ref("posts")
      .once("value")
      .then((snapshot) => {
        const posts = [];
        snapshot.forEach((childSnapshot) => {
          posts.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setPosts(posts));
      });
  };
};
