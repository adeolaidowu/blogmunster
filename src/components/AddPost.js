import React from "react";

const AddPost = () => <div>This is my create post page</div>;

export default AddPost;
