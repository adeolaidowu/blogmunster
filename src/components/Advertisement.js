import React from "react";

const Advertisement = () => (
  <div className="text-center">
    <h2>Advertise with us</h2>
    <img src="https://place-hold.it/200x200" alt="placeholder-img" />
    <p>Call 080-XXX-XXX-36</p>
  </div>
);

export default Advertisement;
