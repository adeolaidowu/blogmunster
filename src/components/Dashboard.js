import React from "react";
import { Link } from "react-router-dom";
import PostList from "./PostList";
import PostFilters from "./PostFilters";
import Sidebar from "./Sidebar";

export const Dashboard = () => (
  <div className="dashboard">
    <div className="dashboard__header">
      <h1>My Dashboard</h1>
      <Link className="button btn-primary" to="/create">
        Create New Post
      </Link>
    </div>
    <PostFilters />
    <div className="dashboard__content">
      <PostList />
    </div>
  </div>
);

export default Dashboard;
