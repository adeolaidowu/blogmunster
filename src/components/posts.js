import { v4 as uuidv4 } from "uuid";

//ADD_POST
export const addPost = ({
  title = "",
  content = "",
  imageLink = "",
  createdAt = 0,
} = {}) => ({
  type: "ADD_POST",
  post: {
    id: uuidv4(),
    title,
    content,
    imageLink,
    createdAt,
  },
});

//EDIT_POST
export const editPost = (id, updates) => ({
  type: "EDIT_POST",
  id,
  updates,
});

//DELETE_POST
export const deletePost = ({ id } = {}) => ({
  type: "DELETE_POST",
  id,
});
