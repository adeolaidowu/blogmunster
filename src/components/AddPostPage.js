import React from "react";
import { connect } from "react-redux";
import PostForm from "./PostForm";
import { startAddPost } from "../actions/posts";

export const AddPostPage = (props) => {
  const onSubmit = (post) => {
    props.startAddPost(post);
    props.history.push("/");
  };
  return (
    <div>
      <h1>Create Blog Post</h1>
      <PostForm onSubmit={onSubmit} />
    </div>
  );
};

const mapDispatchToProps = (dispatch) => ({
  startAddPost: (post) => dispatch(startAddPost(post)),
});

export default connect(undefined, mapDispatchToProps)(AddPostPage);
