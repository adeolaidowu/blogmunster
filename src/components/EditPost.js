import React from "react";
import { connect } from "react-redux";
import PostForm from "./PostForm";
import { editPost, deletePost } from "../actions/posts";

const EditPost = (props) => {
  const onSubmit = (post) => {
    props.dispatch(editPost(props.post.id, post));
    props.history.push("/");
  };
  const onDelete = () => {
    props.dispatch(deletePost({ id: props.post.id }));
    props.history.push("/");
  };
  return (
    <div>
      <PostForm post={props.post} onSubmit={onSubmit} />
      <button onClick={onDelete}>Delete</button>
    </div>
  );
};
const mapStateToProps = (state, props) => ({
  post: state.posts.find((post) => props.match.params.id === post.id),
});

export default connect(mapStateToProps)(EditPost);
