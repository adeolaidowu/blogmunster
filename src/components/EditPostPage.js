import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import PostForm from "./PostForm";
import { startEditPost, startDeletePost } from "../actions/posts";

export const EditPostPage = (props) => {
  const onSubmit = (post) => {
    props.startEditPost(props.post.id, post);
    props.history.push("/dashboard");
  };
  const onDelete = () => {
    props.startDeletePost({ id: props.post.id });
    props.history.push("/dashboard");
  };
  return (
    <div className="page-container">
      <div className="page-header">
        <Link to="/dashboard" className="button">
          Dashboard
        </Link>
        <button className="button btn-danger" onClick={onDelete}>
          Delete Post
        </button>
      </div>
      <h1>Edit Blog Post</h1>
      <PostForm post={props.post} onSubmit={onSubmit} />
    </div>
  );
};
const mapStateToProps = (state, props) => ({
  post: state.posts.find((post) => props.match.params.id === post.id),
});

const mapDispatchToProps = (dispatch) => ({
  startEditPost: (id, updates) => dispatch(startEditPost(id, updates)),
  startDeletePost: (data) => dispatch(startDeletePost(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPostPage);
