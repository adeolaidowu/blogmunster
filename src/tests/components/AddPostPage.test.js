import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import { AddPostPage } from "../../components/AddPostPage";
import posts from "../fixtures/posts";

let addPost, history, wrapper;
beforeEach(() => {
  addPost = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<AddPostPage addPost={addPost} history={history} />);
});

test("should render add post page correctly", () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should correctly handle submit", () => {
  wrapper.find("PostForm").prop("onSubmit")(posts[1]);
  expect(addPost).toHaveBeenLastCalledWith(posts[1]);
  expect(history.push).toHaveBeenLastCalledWith("/");
});
