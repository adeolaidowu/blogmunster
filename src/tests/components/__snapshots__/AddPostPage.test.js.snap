// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render add post page correctly 1`] = `
<div>
  <h1>
    Create Blog Post
  </h1>
  <PostForm
    onSubmit={[Function]}
  />
</div>
`;
